[Lens] Dimension editor design changes

stratoula opened the pull request:
## Summary

Closes https://github.com/elastic/kibana/issues/129261

It implements all the design changes described here https://www.figma.com/file/26yPFDUxarYzAL3JdoJO4l/Mockups-%2F-Version-1-%5BIn-Progress%5D?node-id=0%3A1

<img width="418" alt="image" src="https://user-images.githubusercontent.com/17003240/179544823-192a3efe-99a2-40f5-8d5a-803705d2db5e.png">
<img width="444" alt="image" src="https://user-images.githubusercontent.com/17003240/179544883-6255dd2f-0a83-457c-a72f-7a7a245336b4.png">
<img width="499" alt="image" src="https://user-images.githubusercontent.com/17003240/179544959-8edb7153-5d3e-4906-896f-fa4bfba18796.png">


Things that should be carried to another PR (I will create an issue)
- Move Group by this field first from appearance to the Data section
- Move Collapse by from appearance to the Data section
This is more complex as the logic needs to change here. 

### Checklist

- [x] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [ ] [Documentation](https://www.elastic.co/guide/en/kibana/master/development-documentation.html) was added for features that require explanation or tutorials
- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [x] Any UI touched in this PR is usable by keyboard only (learn more about [keyboard accessibility](https://webaim.org/techniques/keyboard/))
- [x] Any UI touched in this PR does not create any new axe failures (run axe in browser: [FF](https://addons.mozilla.org/en-US/firefox/addon/axe-devtools/), [Chrome](https://chrome.google.com/webstore/detail/axe-web-accessibility-tes/lhdoppojpmngadmnindnejefpokejbdd?hl=en-US))
- [x] This renders correctly on smaller devices using a responsive layout. (You can test this [in your browser](https://www.browserstack.com/guide/responsive-testing-on-local-server))
- [x] This was checked for [cross-browser compatibility](https://www.elastic.co/support/matrix#matrix_browsers)

elasticmachine commented:
Pinging @elastic/kibana-vis-editors @elastic/kibana-vis-editors-external (Team:VisEditors)

flash1293 commented:
Lots of small things, I'm adding checkboxes to the items so we can cross them off once addressed or invalid

* [ ] I think the field should not be rendered in the multi field way as long as only one field is selected in top values. Also, should it have a grey background? It looks a little weird in this design, maybe go with a white background like we do for intervals custom ranges?:
<img width="341" alt="Screenshot 2022-07-21 at 09 16 09" src="https://user-images.githubusercontent.com/1508364/180153186-62c8d256-9565-4fdd-9dd9-7e24181329c0.png">

* [ ] It seems like the spacing between functions and fields is too large:
<img width="348" alt="Screenshot 2022-07-21 at 09 16 53" src="https://user-images.githubusercontent.com/1508364/180153303-0a6ab51d-526b-4b0e-aa2d-0087fa35f5b4.png">

* [ ] "Add field" button is sticking to the bottom of the multi field select
<img width="349" alt="Screenshot 2022-07-21 at 09 17 41" src="https://user-images.githubusercontent.com/1508364/180153408-66a49a9c-2ea3-4203-9854-7fd27884d8e5.png">

* [ ] It seems like the selected function has a slightly lower font weight than other options, did this happen on purpose?
<img width="329" alt="Screenshot 2022-07-21 at 09 18 22" src="https://user-images.githubusercontent.com/1508364/180153555-14c4575e-e53b-49a5-81c3-ffbed69709c9.png">

* [ ] Just a suggestion, but should we add a tooltip to the incompatible indicator dot? I know we didn't have it before, but somehow I found myself hovering over it expecting it to tell me what it does

* [ ] Too much space between custom editor and advanced accordion
<img width="352" alt="Screenshot 2022-07-21 at 09 21 02" src="https://user-images.githubusercontent.com/1508364/180154028-9d13b75b-2bf4-4d29-af9a-32e1b469a44a.png">

* [ ] Formula editor standing out with its sharp edges - can we apply border radius here too?
<img width="343" alt="Screenshot 2022-07-21 at 09 21 53" src="https://user-images.githubusercontent.com/1508364/180154174-69c78856-618b-4fa1-894e-7d34313bd8b2.png">

* [ ] Color palette picker has a grey background - is this on purpose?
<img width="383" alt="Screenshot 2022-07-21 at 09 23 33" src="https://user-images.githubusercontent.com/1508364/180154509-806ee7fb-8211-4ebe-8bdb-4faa5252ba77.png">

* [ ] In the mockups, the unselected item in the button group has a white background, not a grey one
<img width="359" alt="Screenshot 2022-07-21 at 09 25 22" src="https://user-images.githubusercontent.com/1508364/180154852-2bf571dd-1e08-4111-96b2-48cda2681e6f.png">

stratoula commented:
@flash1293 about the multifields. I don't have a strong opinion, I just implemented this comment
<img width="1017" alt="image" src="https://user-images.githubusercontent.com/17003240/180160229-d93d8489-627b-4933-9c7a-3662a02ebf31.png">

flash1293 commented:
Ah right, nvm then

stratoula commented:
- @flash1293 very good points! I implemented everything except from one (see next point). I also changed the background of the terms fields into white as you proposed, I really prefer it that way if @MichaelMarcialis is ok with it
<img width="434" alt="image" src="https://user-images.githubusercontent.com/17003240/180177717-1c1b6e36-3abb-4cd9-ba53-8ba1ed80da92.png">

- About the not selected button on the button group. I dont find any way to do this in EUI. I could do it with custom css and somehow select this button but is this what we want @MichaelMarcialis ?

KOTungseth commented:
UI copy looks good 👍

MichaelMarcialis reviewed (commented):
Thanks for putting this together so quickly, @stratoula! Excellent work here. Comments for my initial review can be found below. Let me know if you have any questions or require additional design support.

## Primary/First Form Section

- What visually appears to be a singular first form section in the configuration flyout is actually comprised of multiple `.lnsIndexPatternDimensionEditor__section` sections and a separate heading in the markup. This amalgamation of separate sections and headers appears to be causing some undesired stylistic issues, particularly in regard to incorrect padding/margining between some containers and their children (as detailed in the comments below). Is there a particular reason for the markup to be structured this way for the first form section, rather than housing everything within a single sectioning element?

![image](https://user-images.githubusercontent.com/3884767/180270542-9f2c3e07-34fc-45b1-85a3-c1bbde8ac99b.png)

- The spacing above the header in the first form section appears to be too small. Could we fix it to be 16px to better match the spacing in other sections and the designs?

![image](https://user-images.githubusercontent.com/3884767/180270814-2ffa31f0-6832-4f45-b70a-42da8b6a2792.png)

- Should we be using an `EuiFormRow` as the parent component for the "Method", "Functions", and "Formula" labels and form elements? Currently, we appear to be using a custom markup structure for each of these.

- The spacing between "Functions" and "Field" form elements appears too large currently. Can we change this to be 16px?

![image](https://user-images.githubusercontent.com/3884767/180271016-7b38cffb-741d-42fe-b5e6-44b4fc982622.png)

- When transitioning from "Formula" back to "Quick functions", the "Formula currently applied" callout appears. Currently, this callout is full width and flush with the edges of the flyout. Can we change this to instead adhere to the padding of the the form section (similar to the rest of the elements in the first form section) and use the default EUI callout styles?

![image](https://user-images.githubusercontent.com/3884767/180271357-74ef7a2b-ad04-4f2b-b3a3-a88f80777547.png)

- I was hoping we could reduce the spacing (to 8px) between the switches and their related/preceding form inputs in the first form section. For example, the space between "Field", "Include empty rows", and "Bind to global time picker" is 8px in the designs. Would it be possible to replicate this spacing with all switches in this first form section?

![image](https://user-images.githubusercontent.com/3884767/180271637-595558d1-a670-49e9-93bf-57fcf732b497.png)

- “Bind to global time picker” should be hidden/removed instead of disabled when the default date field is selected.

### Method Button Groups

- Can we change all "Method" button groups to use `buttonSize="compressed"` to better match the designs?

- Let's change the text of "Quick functions" to be "Quick function".

### Quick Function List Groups

- Can we change the text of the "Functions" label for the functions list group to be "Function"?

- Hovering the currently selected quick function list group item reveals a color mismatch between the text and underline. I'm guessing this is happening due to some custom CSS overriding the selected list group item color to be blue. Instead, can we just run with the default EUI selected styling here, which I believe will end up making the selected text gray? I realize this runs counter to the designs (which display the selected item text as blue), but I think that represents forthcoming changes to the list group component that have not been officially implemented yet and are only being tested in Figma.

![image](https://user-images.githubusercontent.com/3884767/180271858-dbc4fe2c-b297-4ec1-9fe7-61631bbe9520.png)

#### Incompatibility Icon and Tooltip

- Can we add 4px of space between the incompatible dot and its preceding text to better match the designs?

- The incompatible dot icon appears to be using the default icon size. Can we change this to use the small (`s`) size to better match the designs?

- The hope was that we could utilize the `EuiIconTip` component for the incompatible dot, so that it could be accompanied by a tooltip explaining its presence to the user ([example in design](https://www.figma.com/file/26yPFDUxarYzAL3JdoJO4l/Mockups-%2F-Version-1-%5BIn-Progress%5D?node-id=2965%3A108653)). Would it be possible to make this change?

- When transitioning from "Formula" back to "Quick functions", all functions are currently being given the incompatible dot treatment. In this case, I think the provided warning callout is sufficient. Can we remove the incompatible dots in this case, so as to not visually overwhelm the user?

![image](https://user-images.githubusercontent.com/3884767/180272029-162c64fc-bb47-4528-a9dc-d0aa700922ae.png)

### Multiple Fields Interface

- For the multiple "Fields" interface that appears when selecting the "Top value" function, it looks like the entire row is draggable (drag handle, field selector, delete button). Can we change this so that only the drag handle icon can be used to drag, rather than the entire row? I think this will better prevent some cursor collision issues, such as the showing of the grab cursor when hovering the field selector.

- Each multiple "Fields" row's drag handle appears to use the lighter subdued color. Could we change this to instead use the darker default text color to better match the designs?

- When dragging one of the multiple "Fields" rows, the green boundary box currently appears as a separate inset container within the existing gray container. Can we instead change it so that the existing gray container turns green during the drag event (and thus avoiding the box-in-a-box look)?

![image](https://user-images.githubusercontent.com/3884767/180272317-a1e687c2-a3a2-40ca-b4f1-051bddec0ab1.png)

- It appears that the space between the last multiple "Fields" row and the edge of the parent gray container is slightly too large. Can we change the space between the last row and container edge to be exactly 8px? It appears that the added spacing is currently coming from bottom margining on each row.

- The "Add field" button should have 4px of space between itself and the preceding container.

- The "Add field" button should use the `flush="left"` prop so that it aligns nicely with the left edge of the flyout content.

### Formula Editor

- All corners of the "Formula" editor should be rounded (in inline mode, not full screen) to better match the designs.

![image](https://user-images.githubusercontent.com/3884767/180272522-141e3925-f959-4ea3-aac1-7f50fa818c98.png)

- Can we reduce the left/right padding in the "Formula" editor's header and footer toolbars to 8px?

- Is it possible for us to hide the "Formula" editor's scrollbar track until scrolling is needed due to text overflowing?

![image](https://user-images.githubusercontent.com/3884767/180272751-a8e9c240-be37-4829-a3e5-29335cb85e3b.png)

### Advanced Accordion

- The spacing before the "Advanced" accordions and preceding form inputs appears too large in some situations and too small in others. Can we fix this to be a consistent 16px in all cases?

- Can we change the "Advanced" accordion font weight to be bold to better match the designs?

- Hovering the "Advanced" accordion text currently reveals a color mismatch between the text and underline. Can we fix this so the color remains blue throughout?

![image](https://user-images.githubusercontent.com/3884767/180272960-1aad0c60-0ab9-4b21-b52b-f431c3593467.png)

- The switch to "Hide zero values" in the "Advanced" accordion is the wrong font size and weight compared to other switches in first form group and designs. Would it be possible to adjust to match?

![image](https://user-images.githubusercontent.com/3884767/180273076-14cdb2eb-2ca9-4392-b6a6-160a0b409849.png)

## Edit Color Form

- The edit "Color" flyout's form section still has gray background and bottom border. Would it be possible to remove both to better match the designs?

![image](https://user-images.githubusercontent.com/3884767/180273345-b837d4c7-8733-40bd-ac20-a9e1e594b9b0.png)

- While reviewing, I noticed that the edit "Color" flyout's color range swatches are the wrong size. Not sure if this is a separate issue or not, but these should be 32x32px to match the designs. Can we change while we're in here?

- Not sure if a separate issue, but I noticed the edit "Color" flyout's arrow icon the precedes the max/last value input is the wrong size and color. Can we use the default (`m`) EUI icon size and give it the disabled text color?

- Not sure if a separate issue, but I noticed the edit "Color" flyout's actions to "Add color", "Reverse colors", and "Distribute values" are too tightly spaced. Can we change to have 16px of space between the buttons to better match the designs?

## Additional Comments

- The "Table" visualization's metric configuration flyout's summary form inputs weren't moved to their own "Summary" section as suggested in the design. Would it be possible to do?

![image](https://user-images.githubusercontent.com/3884767/180273527-68e44c74-e550-402f-8660-bd7624fbd87e.png)

- Not sure if a separate issue, but I noticed that Annotation configuration flyout's date inputs (for both single and range) should be using compressed form styles. Would that be possible to change while we're in here?

MichaelMarcialis commented:
> * @flash1293 very good points! I implemented everything except from one (see next point). I also changed the background of the terms fields into white as you proposed, I really prefer it that way if @MichaelMarcialis is ok with it

If it's alright with you both, I'd prefer that we keep the gray background on the multiple "Fields" interface. I feel like it does a better job of conveying the group and boundaries of the draggable area (prior to the user actually performing a drag action), while also helping to indicate to the user that additional fields can be added. Also, I suggested in my comments above that it would be nice if we could transition this gray container to the drop zone green color when a drag event is detected (rather than putting a separate green box inside this gray box). I think that will look nice. Thoughts?

> * About the not selected button on the button group. I dont find any way to do this in EUI. I could do it with custom css and somehow select this button but is this what we want @MichaelMarcialis ?

This can be achieved by using the compressed version of the button group in EUI. I left a comment mentioning that above.

stratoula commented:
@MichaelMarcialis thanx for the detailed review and the help here! I will address your comments asap!

stratoula commented:
@flash1293 @MichaelMarcialis I will return this to a draft state, with the current DOM structure there are some limitations as Michael pointed correctly. So I want to look into it thoroughly.

stratoula commented:
@MichaelMarcialis I think I have addressed almost anything. Some comments:
- > For the multiple "Fields" interface that appears when selecting the "Top value" function, it looks like the entire row is draggable (drag handle, field selector, delete button). Can we change this so that only the drag handle icon can be used to drag, rather than the entire row? I think this will better prevent some cursor collision issues, such as the showing of the grab cursor when hovering the field selector.

I couldn't make it work as we want easily. As this behavior exists on other draggable components in Lens (i.e. filters agg on breakdown) I think this should be a separate issue in order to align all of them

- > When dragging one of the multiple "Fields" rows, the green boundary box currently appears as a separate inset container within the existing gray container. Can we instead change it so that the existing gray container turns green during the drag event (and thus avoiding the box-in-a-box look)?

I tried to do it but check the gif here
![lens_multi_terms](https://user-images.githubusercontent.com/17003240/181188049-7bc2f45a-44df-48dc-baf7-4c81f23d6a83.gif)
 
the draggable fields are using the success color with opacity 0.25. I used the same for the container but as there is the opacity I can't make the container bgColor and the draggable bgColor to match. Any ideas how we can achieve that?

- > Is it possible for us to hide the "Formula" editor's scrollbar track until scrolling is needed due to text overflowing?

As I have said on the text-based editor PR this is something that needs to be investigated. I havent found any way to do it with monaco so far. As we are going to investigate it there, if we find a solution we can also do it in Lens editor.

MichaelMarcialis commented on x-pack/plugins/lens/public/indexpattern_datasource/dimension_panel/dimensions_editor_helpers.tsx:
```diff
@@ -73,6 +75,9 @@ export const CalloutWarning = ({
           })}
           iconType="alert"
           color="warning"
+          css={css`
+            margin-bottom: ${euiTheme.size.base};
+          `}
```
It looks like both this and the next callout are using the same exact inline CSS. Is it worth removing those `css` props and instead apply these styles as a single style by targeting the `.lnsIndexPatternDimensionEditor__warning` class selector in `dimension_editor.scss`? Doing so would avoid repeated code and avoid implementing competing style methodologies (old Sass way and new Emotion way), at least until we decide to tackle a transition to Emotion in a more holistic fashion.

stratoula replied:
Got it!

MichaelMarcialis commented on x-pack/plugins/lens/public/indexpattern_datasource/operations/definitions/date_histogram.tsx:
```diff
@@ -264,38 +266,119 @@ export const dateHistogramOperation: OperationDefinition<
 
     return (
       <>
+        <EuiSpacer size="s" />
```
Very clever use of `EuiSpacer` here to prevent the adjacent sibling styles from `EuiFormRow` from being applied!

MichaelMarcialis commented on x-pack/plugins/lens/public/indexpattern_datasource/operations/definitions/date_histogram.tsx:
```diff
@@ -346,81 +429,33 @@ export const dateHistogramOperation: OperationDefinition<
+              'xpack.lens.indexPattern.dateHistogram.dropPartialBucketsHelp',
+              {
+                defaultMessage:
+                  'Drop partial intervas is disabled as these can be computed only for a time field bound to global time picker in the top right.',
```
Typo fix:

```suggestion
                  'Drop partial intervals is disabled as these can be computed only for a time field bound to global time picker in the top right.',
```

MichaelMarcialis commented on x-pack/plugins/lens/public/indexpattern_datasource/operations/definitions/date_histogram.tsx:
```diff
@@ -264,38 +266,119 @@ export const dateHistogramOperation: OperationDefinition<
-          })}
+          helpText={
+            <>
+              <span>
+                {i18n.translate('xpack.lens.indexPattern.dateHistogram.selectOptionHelpText', {
+                  defaultMessage: `Select an option or create a custom value.`,
+                })}
+              </span>
+              <br />
+              <span>
+                {i18n.translate(
+                  'xpack.lens.indexPattern.dateHistogram.selectOptionExamplesHelpText',
+                  {
+                    defaultMessage: `Examples: 30s, 20m, 24h, 2d, 1w, 1M`,
+                  }
+                )}
+              </span>
```
I believe you can remove the `span` elements here.

MichaelMarcialis commented on x-pack/plugins/lens/public/indexpattern_datasource/operations/definitions/formula/editor/formula_editor.tsx:
```diff
@@ -651,7 +656,26 @@ export function FormulaEditor({
+        className="lnsIndexPatternDimensionEditor--shaded"
+        css={css`
+          border: ${!isFullscreen ? euiTheme.border.thin : 'none'};
+          border-radius: ${euiTheme.border.radius.medium};
```
Should this be conditional on fullscreen as well?

```suggestion
          border-radius: ${!isFullscreen ? euiTheme.border.radius.medium : 0};
```

stratoula replied:
Of course!

MichaelMarcialis reviewed (commented):
Thanks for making those requested changes, @stratoula! This is looking great! I've left you some replies and comments below for your re-review. Let me know if you have any questions.

## Previous Comments

> What visually appears to be a singular first form section in the configuration flyout is actually comprised of multiple `.lnsIndexPatternDimensionEditor__section` sections and a separate heading in the markup. This amalgamation of separate sections and headers appears to be causing some undesired stylistic issues, particularly in regard to incorrect padding/margining between some containers and their children (as detailed in the comments below). Is there a particular reason for the markup to be structured this way for the first form section, rather than housing everything within a single sectioning element?

It appears that in the effort to consolidate elements under a unified element, we're now using `EuiFormRow` components as the parent element for each form section. Unfortunately, using `EuiFormRow` this way creates semantic (and likely accessibility) issues. Was there a specific reason for this usage? If not, I'd suggest we change from using `EuiFormRow` as the form sectioning element to either a basic HTML element (ex. `div`, `section`) or perhaps a less situational EUI component such as `EuiPanel`.

> When transitioning from "Formula" back to "Quick functions", the "Formula currently applied" callout appears. Currently, this callout is full width and flush with the edges of the flyout. Can we change this to instead adhere to the padding of the the form section (similar to the rest of the elements in the first form section) and use the default EUI callout styles?

Thanks for making this change. It looks like the last thing needed is some spacing between this callout and the preceding "Method" button group. 8px should do it.

![image](https://user-images.githubusercontent.com/3884767/181369221-39385e7c-144b-4b27-aa98-f931ce8e5f5d.png)

> > For the multiple "Fields" interface that appears when selecting the "Top value" function, it looks like the entire row is draggable (drag handle, field selector, delete button). Can we change this so that only the drag handle icon can be used to drag, rather than the entire row? I think this will better prevent some cursor collision issues, such as the showing of the grab cursor when hovering the field selector.
>
> I couldn't make it work as we want easily. As this behavior exists on other draggable components in Lens (i.e. filters agg on breakdown) I think this should be a separate issue in order to align all of them

Sounds good. Let me know if you want me to make that issue or if you're already on it.

> > When dragging one of the multiple "Fields" rows, the green boundary box currently appears as a separate inset container within the existing gray container. Can we instead change it so that the existing gray container turns green during the drag event (and thus avoiding the box-in-a-box look)?
>
> I tried to do it but check the gif here

Thanks, this looks close! It appears that the containing gray box is being given the green color, but there's still an additional inner box (the drop zone, I presume) that's also being given a green color. Rather than having the two competing containers, can we simply have a single colored drop zone container (gray by default, green during drag event)?

![image](https://user-images.githubusercontent.com/3884767/181369435-d35d0a55-c517-44c1-bcd3-4471f6eef99a.png)

> Hovering the "Advanced" accordion text currently reveals a color mismatch between the text and underline. Can we fix this so the color remains blue throughout?

It seems that this issue has been fixed on hover but still occurs on focus. Would it be possible to fix that as well?

![image](https://user-images.githubusercontent.com/3884767/181369601-3e1fcf17-0f07-47a8-8d79-37e628ad614a.png)

> Not sure if a separate issue, but I noticed the edit "Color" flyout's arrow icon the precedes the max/last value input is the wrong size and color. Can we use the default (`m`) EUI icon size and give it the disabled text color?

Thanks for fixing this. Unfortunately, the change in size has caused the adjacent form input to appear out of alignment with the inputs above. Would it be possible to correct? I imagine this can be dealt with by applying a combination of min-width and center text alignment on the containing `EuiFlexItem`.

![image](https://user-images.githubusercontent.com/3884767/181369804-ac5118b6-0182-4d2e-b8d1-0b1285c977b0.png)

> Not sure if a separate issue, but I noticed that Annotation configuration flyout's date inputs (for both single and range) should be using compressed form styles. Would that be possible to change while we're in here?

It looks like this issue persists.

![image](https://user-images.githubusercontent.com/3884767/181369918-ae9ea0c7-aff0-44dd-8882-576e9bff85f6.png)

## New Comments

- The incompatible dots for quick functions are no longer appearing at all in metric-type dimensions configurations (such as XY visualizations' vertical axes). Not sure if this is due to the previous request to not show the incompatible dots when switching back to quick functions from formula, but wanted to mention it nonetheless.

- The containing element for the top values multiple fields interface gets slightly taller during a drag event. Can we ensure the height of the container remains the same during the drag event?

stratoula commented on packages/kbn-coloring/src/shared_components/coloring/color_ranges/color_ranges_extra_actions.tsx:
```diff
@@ -50,9 +51,18 @@ export function ColorRangesExtraActions({
 
+  const { euiTheme } = useEuiTheme();
+
+  const styles = useMemo(
+    () => css`
+      margin-right: ${euiTheme.size.base};
+    `,
+    [euiTheme.size.base]
+  );
```
Correct!

stratoula commented on x-pack/plugins/lens/public/indexpattern_datasource/dimension_panel/advanced_options.tsx:
```diff
@@ -17,13 +18,19 @@ export function AdvancedOptions(props: { options: AdvancedOption[] }) {
-          {i18n.translate('xpack.lens.indexPattern.advancedSettings', {
-            defaultMessage: 'Advanced',
-          })}
+          <strong>
+            {i18n.translate('xpack.lens.indexPattern.advancedSettings', {
+              defaultMessage: 'Advanced',
+            })}
+          </strong>
```
Ha! Smart! Thanx :)

stratoula commented:
@MichaelMarcialis thanx again for the amazing review. I have addressed almost everything. There is one comment I didn't manage to find a solution (sorry for that)

- > Not sure if a separate issue, but I noticed that Annotation configuration flyout's date inputs (for both single and range) should be using compressed form styles. Would that be possible to change while we're in here?

How can I accomplish that? On the singe date picker the row is already compressed

<img width="939" alt="image" src="https://user-images.githubusercontent.com/17003240/181488611-ddd61b04-f567-44a2-8ef4-bb92cdeb658a.png">


On the range I add compressed to the EuiFormControlLayout but the result is rather disappointing :/

<img width="408" alt="image" src="https://user-images.githubusercontent.com/17003240/181488841-8a2745a4-8b0e-4f12-bf78-f6e9b99f52d1.png">


### Additional comments
- About the wrong semantic issue with the EuiFormRows my bad! I totally misunerstood one one of your comments! It must be ok now (I used a div)
- About the draggable handler, I will create one issue when this PR is merged. Dont worry about it

MichaelMarcialis reviewed (approved):
Thanks again, @stratoula. This looks fantastic! Aside from these last few comments below, this PR looks good from my perspective. Approving now so I don't hold you up further.

> > Not sure if a separate issue, but I noticed that Annotation configuration flyout's date inputs (for both single and range) should be using compressed form styles. Would that be possible to change while we're in here?
>
> How can I accomplish that? On the singe date picker the row is already compressed
>
> On the range I add compressed to the EuiFormControlLayout but the result is rather disappointing :/

I did some looking, and surprisingly it seems that the `EuiDatePicker` component doesn't support a compressed style right now (per issue https://github.com/elastic/eui/issues/4572). Apologies for the runaround there. We can ignore for now until EUI add support.

> The incompatible dots for quick functions are no longer appearing at all in metric-type dimensions configurations (such as XY visualizations' vertical axes). Not sure if this is due to the previous request to not show the incompatible dots when switching back to quick functions from formula, but wanted to mention it nonetheless.

It appears that this issue still persists.

> The containing element for the top values multiple fields interface gets slightly taller during a drag event. Can we ensure the height of the container remains the same during the drag event?

It appears that this issue still persists.

stratoula commented:
> The incompatible dots for quick functions are no longer appearing at all in metric-type dimensions configurations (such as XY visualizations' vertical axes). Not sure if this is due to the previous request to not show the incompatible dots when switching back to quick functions from formula, but wanted to mention it nonetheless.

Discussed with Michael offline, it works correctly!

dej611 commented:
> * > Is it possible for us to hide the "Formula" editor's scrollbar track until scrolling is needed due to text overflowing?
> 
> As I have said on the text-based editor PR this is something that needs to be investigated. I havent found any way to do it with monaco so far. As we are going to investigate it there, if we find a solution we can also do it in Lens editor.

I think the confusing bit is that... that thing is not a scrollbar, rather a ruler: the scrollbar is already set to `auto`, which makes it visible only when required.
I've looked into the monaco API and I think that setting `overviewRulerLanes: 0` in the initial configuration should solve the problem. (Tested on https://microsoft.github.io/monaco-editor/playground.html#customizing-the-appearence-scrollbars )

stratoula commented:
Now I am confused :) Do we want to remove the ruler? Why? 🤔

dej611 commented:
> Now I am confused :) Do we want to remove the ruler? Why? 🤔

Just thought that was @MichaelMarcialis was asking.

stratoula commented:
Got it, nevertheless let's do it on another PR! I will talk with Michael to understand precisely what he wants.

MichaelMarcialis commented:
> I have only a minor comment for the `Count of Records` state now: the layout looks a bit more "noisy" with all these yellow dots compared to the previous version.

@dej611: Thanks for the feedback. I agree that the list of quick functions can appear a bit "noisy" when the count function is selected. It's something I debated with myself on quite a bit. Ultimately, I felt that it was more important to make it clear to the user that the incompatible functions were still available and not disabled, at the cost of a little more visual noise.

That said, as we continue to add quick functions over time, this list of functions continues to grow larger and takes up valuable vertical real estate in the configuration flyout. Because of this, I imagine we are going to hit a threshold at which we'll need to reevaluate our use of `EuiListGroup` here and perhaps switch it out for another, more scalable and compact component. That will likely also have the side benefit of dramatically reducing the visual noise from showing many incompatible functions.

> > Now I am confused :) Do we want to remove the ruler? Why? 🤔
> 
> Just thought that was @MichaelMarcialis was asking.

Apologies for the confusion. This is likely due to my not understanding the intricacies of the Monaco editor. So this "ruler" (which shows errors, current cursor position, etc.) is different/independent from the scroll track in the Monaco editor and they are just visually overlapping each other? If that is indeed the case, let's just ignore my request to remove what I previously thought was only the scroll track and run with the defaults (assuming it's consistent with what we do with the Monaco editor elsewhere in Kibana).
